src/rtps/main.cpp

```diff
@@ -0,0 +1,289 @@
+                "  -d <device>              UART device to get RTPS (and MAVLink) messages.\n"
+                "                           Default '/dev/ttyS1'\n"
+                "  -b <baudrate>            Baudrate of UART device.\n"
+                "                           Default '460800'\n"
+                "  -p <port>                UDP port where to send (or receive) MAVLink packets.\n"
+                "                           Default '13800'\n"
+                "  -r                       Report stats each 10 seconds\n",
+            program_invocation_short_name);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+    int c;
+
+    while ((c = getopt(argc, argv, "d:b:p:r")) >= 0) {
+        switch (c) {
+        case 'd':
+            _options.device = optarg;
+            break;
+        case 'b':
+            if (safe_atoul(optarg, &_options.baudrate)) {
+                log_error("Invalid argument for baudrate = %s", optarg);
+                help(stderr);
+                return -EINVAL;
+            }
+            break;
+        case 'p':
+            if (safe_atoul(optarg, &_options.port)) {
+                log_error("Invalid argument for port = %s", optarg);
+                help(stderr);
+                return -EINVAL;
+            }
+            break;
+        case 'r':
+            _options.report_stats = true;
+            break;
+        case '?':
+        default:
+            help(stderr);
+            return -EINVAL;
+        }
+    }
+
+    return 0;
+}
+
+static int setup_udp(unsigned port)
+{
+    int fd;
+    const int broadcast_val = 1;
+
+    fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (fd == -1) {
+        log_error("Could not create socket %d", errno);
+        return -1;
+    }
+
+    _sockaddr.sin_family = AF_INET;
+    _sockaddr.sin_addr.s_addr = inet_addr("127.0.0.1");
+    _sockaddr.sin_port = htons(port);
+
+    if (setsockopt(fd, SOL_SOCKET, SO_BROADCAST, &broadcast_val, sizeof(broadcast_val))) {
+        log_error("Error enabling broadcast %d", errno);
+        return -1;
+    }
+
+    if (fcntl(fd, F_SETFL, O_NONBLOCK | FASYNC) < 0) {
+        log_error("Error setting socket fd as non-blocking %d", errno);
+        return -1;
+    }
+
+    return fd;
+}
+
+static void loop(int epollfd, int uart_fd, int udp_fd, UART_node uart_node, RtpsTopics &topics)
+{
+    unsigned bytes_read = 0, bytes_sent = 0;
+    char topic_ID = 255;
+
+    struct timespec begin;
+    bool measuring = false;
+    uint8_t seq;
+
+    struct epoll_event events[8];
+    char data_buffer[BUFFER_SIZE] = {};
+
+    // TODO use mavlink-router mainloop (which need to be extracted from it)
+    do {
+        if (_options.report_stats && !measuring) {
+            clock_gettime(0, &begin);
+            measuring = true;
+        }
+
+        // TODO check if it's possible to add RTPS subscribers on epoll
+        // or use a thread for them
+        int r = epoll_wait(epollfd, events, 8, USLEEP_TIME);
```

Comment:
You kill throughout when you keep waking to do nothing.  This timeout should be used to avoid locking the main loop when something really bad happens, not to wakeup periodically. And if you are not handling the case in which you wakeup and have no FD ready, there's no point to wakeup: it should rather be a blocking wait.